feat: ✨ add flag to disable getSession after signIn on local / refresh provider

bitfactory-frank-spee opened the pull request:
<!---
☝️ PR title should follow conventional commits (https://conventionalcommits.org)
-->

### 🔗 Linked issue

<!-- Please ensure there is an open issue and mention its number as #123 -->
https://github.com/sidebase/nuxt-auth/issues/701

### ❓ Type of change

<!-- What types of changes does your code introduce? Put an `x` in all the boxes that apply. -->

- [x] 📖 Documentation (updates to the documentation, readme or JSdoc annotations)
- [x] 🐞 Bug fix (a non-breaking change that fixes an issue)
- [x] 👌 Enhancement (improving an existing functionality like performance)
- [x] ✨ New feature (a non-breaking change that adds functionality)
- [ ] 🧹 Chore (updates to the build process or auxiliary tools and libraries)
- [ ] ⚠️ Breaking change (fix or feature that would cause existing functionality to change)

### 📚 Description

See issue for clarification. This adds a flag to the signInOptions named `withSession` which defaults to true. This makes sure this will be backwards compatible. When `withSession === false` the `getSession` method will not be called.

This also includes a bug fix (681fdb07218397fbe52bf097b6daf044762970a8) where refresh was not equal to local provider.

### 📝 Checklist

- [x] I have linked an issue or discussion.
- [ ] I have added tests (if possible).
- [x] I have updated the documentation accordingly.

NOTE: this documentation mentioned `withGetSession` before and not `withSession` as decided to be the name later.

phoenix-ru commented:
Hi @bitfactory-frank-spee , thanks for your contribution, it looks good to me.

I will come back to it after #694 , so that we can ensure that it's functionally good.

phoenix-ru commented:
@bitfactory-frank-spee Could you please sync your PR with the current main?

phoenix-ru commented:
bump @bitfactory-frank-spee

bitfactory-frank-spee commented:
@phoenix-ru I will update it this week 👍

bitfactory-frank-spee commented:
> @bitfactory-frank-spee Could you please sync your PR with the current main?

Done 🚀

zoey-kaiser commented on docs/content/3.application-side/2.session-access-and-management.md:
```diff
@@ -103,12 +105,16 @@ await signIn(credentials, { callbackUrl: '/protected' })
 await signIn(credentials, { callbackUrl: 'https://sidebase.io', external: true })
 
+// Trigger a sign-in without calling getSession directly. You have to manually call it to get session data.
+await signIn(credentials, { withGetSession: false })
```
Can we rename this new option from `withGetSession` to just `getSession`? I think this would be a bit more readable!

bitfactory-frank-spee replied:
Hi Zoey, you just caught me on holiday 🌴. I will change it asap, next week for certain. I will try to do it sooner. 

Renaming is fine by me. I don't totally agree with the name change. Because it is a boolean you want to name it like `is...`, `has...`, `with...` or something like that, and not a get/set/action like method name. 

So I would suggest just `withSession` or without 'with', e.g. something like `callSession` or just `session`. But again all names are fine by me. Just giving my opinion to get it as good as possible.

zoey-kaiser replied:
Agreed! Let's go with `withSession`

bitfactory-frank-spee replied:
Done in [`ac9415f` (#702)](https://github.com/sidebase/nuxt-auth/pull/702/commits/ac9415fb6ff552478872c391badbdd1d5bf11d86)

zoey-kaiser reviewed (changes requested):
Hi @bitfactory-frank-spee 👋 

Very close to being done! I have one minor renaming request and then we should be good to go, to release this in 0.8!

zoey-kaiser commented:
Hi @bitfactory-frank-spee 👋 

We just released 0.8.0, so I would begin looking into getting this PR merged in the next release. Could you please merge main and we can then do a final round of reviews 😊

zoey-kaiser commented on src/runtime/composables/local/useAuth.ts:
```diff
@@ -32,9 +32,12 @@ const signIn: SignInFunc<Credentials, any> = async (credentials, signInOptions,
-  await nextTick(getSession)
+  const { callbackUrl, redirect = true, external, withSession = true } = signInOptions ?? {}
+
+  if (withSession) {
```
Could you please add type support for the new `withSession` option? Please also consider, that you will need to account for the fact that this value is only available for the `local` and `refresh` providers and should therefore not be added to the signIn function of the `authjs` provider.

bitfactory-frank-spee replied:
Hi Zoey. Sorry it took a while. In my mind I responded somewhere. I can add typing.

But can you direct / help me to how I do this "account for the fact that this value is only available for the local and refresh providers and should therefore not be added to the signIn function of the authjs provider"? Because I have no idea what you mean here.

bitfactory-frank-spee replied:
So where do I add the typing support?

bitfactory-frank-spee replied:
And how do I make sure this is not added to the signIn function of the authjs provider?

bitfactory-frank-spee replied:
Is this enough? https://github.com/sidebase/nuxt-auth/pull/702/files#diff-23ace85c1398da87fad352ed08fd7026dab0a96402ccc8b6c3e23b478c993f43

bitfactory-frank-spee replied:
@zoey-kaiser ?

zoey-kaiser reviewed (changes requested):
Just some typings missing 😊

Mucaccino commented:
I need this. Really this was gone stale?

bitfactory-frank-spee commented:
> I need this. Really this was gone stale?

Ahh, I now see that I needed to respond to this, and that I didn't. I have now: https://github.com/sidebase/nuxt-auth/pull/702#discussion_r1810021411

bitfactory-frank-spee commented:
I merged the current main into this PR and fixed the conflicts as the local and refresh provider are merged in the new release. I also fixed the documentation.

bitfactory-frank-spee commented on src/runtime/types.ts:
```diff
@@ -548,6 +548,12 @@ export interface SecondarySignInOptions extends Record<string, unknown> {
+  withSession?: boolean
 }
 
 export interface SignUpOptions extends SecondarySignInOptions {
```
This interface is odd IMHO. Because I never see this used in the code with any of the properties of the `SecondarySignInOptions` interface. It is just used next to a property which uses `SecondarySignInOptions`. But I could miss something which is there somewhere.

phoenix-ru replied:
You're actually right, it's a weird implementation in https://github.com/sidebase/nuxt-auth/pull/615 which accepts two identical interfaces:
https://github.com/sidebase/nuxt-auth/blob/218a8465fd01fbf70562ab4cc1083dea3931d1a1/src/runtime/composables/local/useAuth.ts#L161

It should be just `signUpOptions`, will probably move this to a different breaking change

phoenix-ru replied:
Issue created: https://github.com/sidebase/nuxt-auth/issues/948

bitfactory-frank-spee replied:
Excellent thanks 👍

phoenix-ru commented on docs/guide/application-side/session-access.md:
```diff
@@ -102,7 +102,7 @@ Time at which the session was last refreshed, either `undefined` if no refresh w
 Returns the current Cross Site Request Forgery Token (CSRF Token) required to make POST requests (e.g. for signing in and signing out).
 
-You likely only need to use this if you are not using the built-in `signIn()` and `signOut()` methods. Read more: https://next-auth.js.org/getting-started/client#getcsrftoken
+You likely only need to use this if you are not using the built-in `signIn()` and `signOut()` methods. Read more: <https://next-auth.js.org/getting-started/client#getcsrftoken>
```
Why did you change the link format? This is a markdown file, and links are correctly recognized by Vitepress
```suggestion
You likely only need to use this if you are not using the built-in `signIn()` and `signOut()` methods. Read more: https://next-auth.js.org/getting-started/client#getcsrftoken
```

bitfactory-frank-spee replied:
Ahh, that is probably markdownlint. I highly recommend to implement this. But I will remove them for now.

bitfactory-frank-spee replied:
It is this rule: https://github.com/DavidAnson/markdownlint/blob/v0.36.1/doc/md034.md

bitfactory-frank-spee replied:
https://github.com/sidebase/nuxt-auth/pull/702/commits/7739fbc1d06753c7f69649aab4011347fa6a43a2

phoenix-ru reviewed (commented):
Hi @bitfactory-frank-spee , sorry for us not answering for a long time, I would take over the PR and make sure it's over the line :)

phoenix-ru commented:
@bitfactory-frank-spee, unfortunately, I cannot directly commit to your PR (probably due to restrictions on your org). I would have to create a branch off your PR to proceed

phoenix-ru commented on src/runtime/types.ts:
```diff
@@ -548,6 +548,12 @@ export interface SecondarySignInOptions extends Record<string, unknown> {
    * @default false
    */
   external?: boolean
+  /**
+   * Whether `getSession` is called. If set to false, you have to manually call `getSession` to get the session data.
+   *
+   * @default true
+   */
```
```suggestion
  /**
   * Whether `getSession` needs to be called after a successful sign-in. When set to false, you can manually call `getSession` to obtain the session data.
   *
   * @default true
   */
```

phoenix-ru commented on src/runtime/types.ts:
```diff
@@ -548,6 +548,12 @@ export interface SecondarySignInOptions extends Record<string, unknown> {
+   *
+   * @default true
+   */
+  withSession?: boolean
```
I am actually not in favor of this name, as I used it when testing your PR and it doesn't "click". Existing options use the imperative forms and I think we should stick to it for the sake of consistency (e.g. `redirect`, `preventLoginFlow`).

Another problem I see with `withSession` is that it may confuse users as to what it means. I propose a more explicit

```suggestion
  callGetSession?: boolean
```

bitfactory-frank-spee replied:
There was an earlier thread with Zoey about this. We ended up with `withSession`. I think it's fine to change, I will do it. See https://github.com/sidebase/nuxt-auth/pull/702/files/e59026bf2b8492e382b04ba8f1f7e12bf072f601#diff-dd4f118f099f41a809eb53d231259e96a9818b3d7f734b4b671de323ac718685

bitfactory-frank-spee commented:
> Hi @bitfactory-frank-spee , sorry for us not answering for a long time, I would take over the PR and make sure it's over the line :)

No problem. I will check your comments today.

bitfactory-frank-spee commented:
> @bitfactory-frank-spee, unfortunately, I cannot directly commit to your PR (probably due to restrictions on your org).
> 
> I still checked it though and it seems to work, but I have several points

Ah that could be. I am sorry about that. I can probably not change that immediately. Next time I will use my personal account.

bitfactory-frank-spee commented:
@phoenix-ru I fixed everything you asked. And updated this branch with the main branch.

bitfactory-frank-spee commented:
> Hi @bitfactory-frank-spee , extremely sorry for deprioritizing this for a long time! I will ask @zoey-kaiser if she's fine with squeezing this PR into the upcoming `0.10.0` release and merge asap

Excellent! 🎉 🚀 Thanks for the help. Good luck with further releases.
